src/main/java/service/StudentService.java

```diff
@@ -0,0 +1,29 @@
+package service;
+
+import entity.Student;
+
+import java.util.List;
+import java.util.Objects;
+
+public class StudentService {
+
+    public Double averageRatingForSubject(List<Student> students, String subjectName) {
+        return students.stream()
+                .map(Student::getRating)
+                .mapToDouble(d -> d.get(subjectName))
+                .average()
+                .orElseThrow(() -> new RuntimeException("Subject not found"));
+    }
+
+    public void printStudentsAndSubjects(List<Student> students) {
+        students.stream()
+                .map(Student::getRating)
+                .flatMap(stringIntegerMap -> stringIntegerMap.keySet().stream())
+                .distinct()
+                .peek(subject -> System.out.print("\n" + subject + " = "))
+                .forEach(subject -> students.stream()
+                        .filter(subj -> Objects.nonNull(subj.getRating().get(subject)))
+                        .forEach(student -> System.out.print(student.getName() + ": "
```

Comment:
This could be more readable if you refactor it in the following way:

create a separate method for this logic and in this method initialize variables that you use in the Sout.